File: 868.二进制间距.py
#
# @lc app=leetcode.cn id=868 lang=python3
#
# [868] 二进制间距
#

# @lc code=start
class Solution:
    def binaryGap(self, N: int) -> int:
        s = str(bin(N))[2:]
        i = 0
        max_dis = 0
        count = 0
        for j in range(1, len(s)):
            if s[j] == '1':
                count += 1
                i = j
                max_dis = max(max_dis, count)
                count = 0
            else:
                count += 1
        return max_dis
        
# @lc code=end

File: 868.二进制间距.cpp
/*
 * @lc app=leetcode.cn id=868 lang=cpp
 *
 * [868] 二进制间距
 */

// @lc code=start
class Solution {
public:
    int binaryGap(int n) {
        string res = "";
        int max_interval = 0;
        int fast = 0, slow = 0;
        while(n){
            if(n%2 == 1)
            {
                if(res[slow] == '1')
                    max_interval = max(max_interval, fast-slow);
                slow = fast;
            }
            res += to_string(n%2);
            n /= 2;
            fast++;
        }
        return max_interval;
    }
};
// @lc code=end

